### ISSUE TITLE
iml and Deep Learning in Keras

### ISSUE THREAD

**Issue by viniciusbrei** [christophM/iml#69]
Hi Chistoph,

May I use iml to get Shapley values from a deep learning model for classification (0 or 1), developed using Keras in R?

I ran the code bellow:

X <- df[which(names(df) != "chg")] # this is my classification variable (0 or 1)
predictor <- Predictor$new(model_keras, data = X, y = df$chg) #model_keras is the keras model object
imp <- FeatureImp$new(predictor, loss = 'ce')

I always get the following error message: "Error: Data passed to Keras must be a vector, matrix, or array (you passed a data frame)".

I converted my data to matrix, but the problem continues. 

Thanks!

**Comment by christophM** [comment]
You have to write a predict.fun, see examples https://github.com/christophM/iml/issues/29

**Comment by viniciusbrei** [comment]
Christoph,
Thank you for your answer. However, #29 is about Xgboost models and my doubt is about deep learning models. Instead of creating the "predictor" object, should I create the "predict.fun"? 
Sorry to ask again, but example #29 was not enough for me to understand what to do. This is the code I tried:

X <- df[which(names(df) != "chg")] # this is my classification variable (0 or 1)
predict.fun=function(model_keras, test_df){    
  newData_x = model_keras(data.matrix(test_df), missing = NA)
  results<-predict(model, newData_x)
  return(results)
}
imp <- FeatureImp$new(predict.fun, loss = 'ce')

error message: Error in predictor$data : object of type 'closure' is not subsettable

**Comment by christophM** [comment]
You also have to create a Predictor object and give that to FeatureImp, see ?Predictor. 
Predictor is also where you use predict.fun as argument.

**Comment by viniciusbrei** [comment]
I am sorry again, but it is still not working. I tried to include the Predictor, and also study your (great) tutorial https://hub.mybinder.org/user/christophm-iml-xsr7fw04/notebooks/notebooks/tutorial-intro.ipynb and replicate the instructions, adapting for a Keras DL model. This is the code I tried:

X <- df[which(names(df) != "chg")] # "chg" is my classification variable (0 or 1)
predictor <- Predictor$new(model_keras, data = X, y = df$chg,
                                                predict.fun=function(object, newdata){
                                                newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)
                                                results<-predict(model, newData_x)
                                                return(results)
                                                }
                                              )
imp = FeatureImp$new(predictor, loss = 'ce')

Error message:  Error in predict(model, newData_x) : object 'model' not found 

Doubts:
1. Is it correct to use "model_keras" (my Keras model object) in Predictor$new(model_keras... ?
2. With which object from Keras do I replace the "xgb.DMatrix" in "newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)"?
3. Should I replace "newdata" and "newData_x" with my own data or should I leave the function the way it is?
4. Should I replace "model" in "results<-predict(model, newData_x)" with my Keras object?

Thank you again!

**Comment by MohGhaziAlZeyadi** [comment]
Hi viniciusbrei, 
I have the same problem, I hope there will be some help soon. 
I follow the example in here ( https://www.exago.ml/blog/2019/4/28/model-explainability-with-keras-and-iml ), but still the same error. 
Thank you

**Comment by simon-tietze** [comment]
Hi MohGhaziAlZeyadi,

which version of iml are you using? Is it the current head from github?

Cheers,
Simon

**Comment by MohGhaziAlZeyadi** [comment]
Hi Simon,
Thank you for your replay, 
actually, I have installed iml 090 and I have used two way to install iml 
# Stable version ------> install.packages("iml")
# Development version------>  devtools::install_github("christophM/iml")
and with both I have got this error 
Error: Data passed to Keras must be a vector, matrix, or array (you passed a data frame)
cheers 
Moh

**Comment by simon-tietze** [comment]
Ok, your version should be good then. I wonder if you are using a different
keras version from the one I used to develop the keras detection.

Could you show me the class of your trained keras model that you pass to
iml?

E.g. class(model)

Cheers,
Simon

**Comment by MohGhaziAlZeyadi** [comment]
Hi Simon,
The class(model) as fallowing:
[1] "keras.engine.sequential.Sequential"                               
[2] "keras.engine.training.Model"                                      
[3] "keras.engine.network.Network"                                     
[4] "keras.engine.base_layer.Layer"                                    
[5] "tensorflow.python.training.checkpointable.base.CheckpointableBase"
[6] "python.builtin.object" 
cheers 
Moh

**Comment by SidaJ** [comment]
I encounter the same issues with multinomial classficiation using #54  (IML + XBG), here is my function hopefully works for you:
predict.fun <- function(model, newdata){
  newData_x = xgb.DMatrix(data.matrix(newdata), missing = NA)
  xgb.pred = predict(model,newData_x,reshape=T)
  xgb.pred = as.data.frame(xgb.pred)
  colnames(xgb.pred) = levels(newdata$label) ##here you need to replace with your Y or Label
  xgb.pred$prediction = apply(xgb.pred,1,function(x) colnames(xgb.pred)[which.max(x)])
  return(xgb.pred$prediction)
}
